app/routines/generate_reset_code.rb

```diff
@@ -0,0 +1,18 @@
+class GenerateResetCode
+
+  lev_routine
+
+  protected
+
+  def exec(identity, expiration_period = ResetCode::DEFAULT_EXPIRATION_PERIOD)
+    identity.reset_code ||= identity.build_reset_code
```

Comment:
The reset code contains the foreign key, so saving the identity is not needed (I set identity.reset_code just so the active record and the database are consistent without reloading the identity).